### DISCUSSION TITLE
problems with custom alist entry

### DISCUSSION THREAD

**Discussion by Seme4eg** [magit/forge#432] [Support] [answered]
config.el:
```elisp
(after! forge
  (push '("gitlab.mycompany.ru" "gitlab.mycompany.ru/api/v4"
          "gitlab.mycompany.ru" forge-gitlab-repository) forge-alist))
```

.authinfo.gpg: `machine gitlab.mycompany.ru/api/v4 login my.username^forge password TOKEN`

.git/config:
```
...
[gitlab "gitlab.mycompany.ru/api/v4"]
	user = my.username
```

added repo to database, fetched issues and PRs, looked like everything's fine, but when i tried to review-pull-request it prompted me with `Git variable 'github.gitlab.mycompany.ru/api/v4.user' is unset. Set to:`

And when I set it (locally) , next thing i get is this 
![image](https://user-images.githubusercontent.com/24554850/140027005-d9046490-49b7-41c1-b8d4-0914222225f0.png)

And another weird thing that for some reason isn't working for me - i don't see commits in PR
![image](https://user-images.githubusercontent.com/24554850/140027120-d535ebfe-8289-4c80-b26f-770b7d2a04c1.png)

**Comment by tarsius** [comment, marked as answer]
Sounds like you need to set the git variable `forge.remote`, see https://magit.vc/manual/forge/Repository-Detection.html.

**Reply by Seme4eg** [reply]
![image](https://user-images.githubusercontent.com/24554850/140165669-95ff222e-42f5-4ce6-bb13-f6302eff7560.png)
it does recognize it
even after i've runned the command given in the link u provided (`git config --local forge.remote my-upstream`) and restarted emacs it shows same error.
in my .git/config i now have 
```
[forge]
	remote = help
```
which is correct remote.

**Reply by tarsius** [reply]
Remove the repository from the database using `M-x forge-remove-repository` and then try adding it again using `N a`.

**Reply by Seme4eg** [reply]
didn't help, it still pulls all PRs and issues, but doesn't show commits in PRs and doesn't allow me to review a PR outputting the same error
don't you find it weird that when asking for my user name it prompts me with `Git variable 'github.gitlab...`? not 'gitlab.gitlab' but 'GITHUB.gitlab'. Even tho i have a gitlab domain. And in .git/config i have this
![image](https://user-images.githubusercontent.com/24554850/140181720-22b472ba-45f7-4dbb-bcb9-20237228c9a8.png)

**Reply by tarsius** [reply]
Yes and it means that Forge thinks it is talking to a Github instance. It's attempt to use GraphQL leads to the same conclusion because we it doesn't use GraphQL when talking to Gitlab, only when talking to Github. I thought of two reasons why that might be (1) `origin` references a repository on Github and Forge wasn't told to use the other remote (2) simply fixing the former is not enough because some wrong data is still stored in the database so the borked information has to be purged. We have already done these things.

This and similar issues concerning corporate Gitlab instances are very painful for me because users seem to never want to disclose their employer and thus obfuscate all the relevant data. I have to poke around in the dark and never know whether an apparent inconsistency in the presented configuration is due to mistakes made during obfuscation or whether they also exist in the actual configuration. And vice-versa, just because something is correct in the obfuscated configuration doesn't mean that it is also correct in the actual configuration. And then of course it is also possible that something is wrong in Forge but I have no way of testing because I have no access to the corporate Gitlab instance, so basically I can just make educated guesses and hope I am guiding the user in the right direction.

**Reply by Seme4eg** [reply]
Totally understandable. Im also upset that it is not my repo and i can't jus give u an access, but it is corporate, rip. Well then, if u will ever come with any idea what else might be the reason or if u need any more data on this topic - i will be here)

**Reply by tarsius** [reply]
I recommend you try to configure Forge for some public third-party Gitlab instance. Maybe the same thing happens when you do that and then you can actually show me.

**Reply by Seme4eg** [reply]
ok, i did try it with another (my own now) gitHUB repo and yes, got same error. And also weird (or maybe it should be like that) thing - i don't see any 'review pull req' shortcut now in this project after i press '@' prefix key in magit buffer: 
![image](https://user-images.githubusercontent.com/24554850/140562023-de707d93-742f-42d6-82ab-fa8052c0a6c1.png)

I did get everything to work again, but reviewing stuff doesn't work again.

config.el -- `machine api.github.com login Seme4eg^forge password TOKEN`
.git/config:
```
[core]
	repositoryformatversion = 0
	filemode = true
	bare = false
	logallrefupdates = true
[remote "origin"]
	url = git@github.com:Seme4eg/.doom.d.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
	remote = origin
	merge = refs/heads/master
	pushRemote = origin
[github]
	user = Seme4eg
[branch "forge-test"]
	pushRemote = origin
```

when i call `github-review-forge-pr-at-point`:
![image](https://user-images.githubusercontent.com/24554850/140562353-a27f1da7-2d4c-486c-9d9b-583cc6ce905a.png)

**Reply by tarsius** [reply]
> when i call `github-review-forge-pr-at-point`

Please ask the author of `github-review` for help.

**Reply by tarsius** [reply]
Actually this sounds familiar, see #380.

**Reply by Seme4eg** [reply]
That did help for my opened github repo, but didn't help for the corporate gitlab one. I did add `my.username^github-review` entry to .gpg file. Now when trying to call review PR i get this error output:
```
(fn &rest ARGS2)] github-review-errback nil nil))
Error talking to GitHub: ((error http 404 ((error . 404 Not Found))) ((Server .
nginx) (Date . Sat, 06 Nov 2021 06:53:51 GMT) (Content-Type . application/json)
(Content-Length . 25) (Connection . keep-alive) (Cache-Control . no-cache) (Vary
. Origin) (X-Content-Type-Options . nosniff) (X-Frame-Options . SAMEORIGIN)
(X-Request-Id . 01FKT0CR2772SR8MR77TYWC667) (X-Runtime . 0.016707)) (:error
(error http 404 ((error . 404 Not Found))) :peer (:certificates ((:version 3
:serial-number 29:c7:88:d4:6f:6c:73:c5:42:53:da:2c :issuer C=BE,O=GlobalSign
nv-sa,CN=AlphaSSL CA - SHA256 - G2 :valid-from 2021-08-17 :valid-to 2022-09-18
:subject CN=*.mycompany.ru :public-key-algorithm RSA
:certificate-security-level High :signature-algorithm RSA-SHA256 :public-key-id
sha1:82:1c:fd:be:32:4d:6f:bd:dd:e1:b8:60:07:61:5a:df:db:32:4b:89 :certificate-id
sha1:09:8f:6e:2b:b5:95:c5:ed:03:30:72:20:c8:6e:7f:5a:90:46:3a:bc :pem -----BEGIN
CERTIFICATE----- 
...
-----END CERTIFICATE-----
) (:version 3 :serial-number 04:00:00:00:00:01:44:4e:f0:36:31 :issuer
C=BE,O=GlobalSign nv-sa,OU=Root CA,CN=GlobalSign Root CA :valid-from 2014-02-20
:valid-to 2024-02-20 :subject C=BE,O=GlobalSign nv-sa,CN=AlphaSSL CA - SHA256 -
G2 :public-key-algorithm RSA :certificate-security-level Medium
:signature-algorithm RSA-SHA256 :public-key-id
sha1:ac:c1:aa:cb:6a:74:41:0a:72:9a:b5:e2:3f:8c:b4:61:55:cd:94:fa :certificate-id
sha1:4c:27:43:17:17:56:5a:3a:07:f3:e6:d0:03:2c:42:58:94:9c:f9:ec :pem -----BEGIN
CERTIFICATE----- 
...
-----END CERTIFICATE-----
)) :certificate (:version 3 :serial-number 29:c7:88:d4:6f:6c:73:c5:42:53:da:2c
:issuer C=BE,O=GlobalSign nv-sa,CN=AlphaSSL CA - SHA256 - G2 :valid-from
2021-08-17 :valid-to 2022-09-18 :subject CN=*.mycompany.ru :public-key-algorithm
RSA :certificate-security-level High :signature-algorithm RSA-SHA256
:public-key-id sha1:82:1c:fd:be:32:4d:6f:bd:dd:e1:b8:60:07:61:5a:df:db:32:4b:89
:certificate-id sha1:09:8f:6e:2b:b5:95:c5:ed:03:30:72:20:c8:6e:7f:5a:90:46:3a:bc
:pem -----BEGIN CERTIFICATE----- 
...
-----END CERTIFICATE-----

(fn D V &rest _)] (#s(deferred identity deferred:default-errorback
deferred:default-cancel #s(deferred #[257
\302\303B\302\242B\240\210\301\302\242GU\203\0\304\300\305\302\242\237#\210\306\207
[#s(deferred deferred:default-callback deferred:default-errorback
deferred:default-cancel #s(deferred deferred:parallel-alist-to-array
deferred:default-errorback #[257 \301!\210\302\300!\207 [#9
deferred:default-cancel deferred:cancel] 3 

(fn X)] #s(deferred #[257
\211A@\301\302\303\304\305\306\236A\236A\236A\236A\236A\236A\304\305\306\236A\236A\236A\307\310\300\311#\312@""\266\202\207
[((owner . v3) (repo . web-arm) (apihost . gitlab.mycompany.ru/api/v4) (num .
372)) oid target headRef pullRequest repository data github-review-save-diff
a-assoc sha github-review-format-diff] 9 

(fn X)] deferred:default-errorback deferred:default-cancel #s(deferred
deferred:default-callback #[257 \300\301"\207 [message Got an error from the
GitHub API %S!] 4 

(fn ERR)] deferred:default-cancel nil nil nil) nil nil) nil nil) nil nil) 2
(nil) 1 deferred:post-task ok nil] 5 

(fn X)] deferred:default-errorback deferred:default-cancel #s(deferred
deferred:default-callback #[257
\302\303B\302\242B\240\210\302\242G\301U\203\0\304\300\305\302\242\237#\210\306\207
[#s(deferred deferred:default-callback deferred:default-errorback
deferred:default-cancel #s(deferred deferred:parallel-alist-to-array
deferred:default-errorback #[257 \301!\210\302\300!\207 [#10
deferred:default-cancel deferred:cancel] 3 

(fn X)] #s(deferred #[257
\211A@\301\302\303\304\305\306\236A\236A\236A\236A\236A\236A\304\305\306\236A\236A\236A\307\310\300\311#\312@""\266\202\207
[((owner . v3) (repo . web-arm) (apihost . gitlab.mycompany.ru/api/v4) (num .
372)) oid target headRef pullRequest repository data github-review-save-diff
a-assoc sha github-review-format-diff] 9 

(fn X)] deferred:default-errorback deferred:default-cancel #s(deferred
deferred:default-callback #[257 \300\301"\207 [message Got an error from the
GitHub API %S!] 4 

(fn ERR)] deferred:default-cancel nil nil nil) nil nil) nil nil) nil nil) 2
(nil) 1 deferred:post-task ok nil] 5 

(fn E)] deferred:default-cancel nil nil nil) nil nil) nil nil)) apply append] 6 

(fn &rest ARGS2)] github-review-errback nil nil))
```

**Reply by tarsius** [reply]
Looks like **github**-review does not support **gitlab**.
